harness/determined/cli/proxy.py

```diff
@@ -145,8 +146,21 @@ def http_connect_tunnel(
 ) -> None:
     parsed_master = request.parse_master_address(master)
     assert parsed_master.hostname is not None, "Failed to parse master address: {}".format(master)
+
+    # The "lomond.WebSocket()" function does not honor the "no_proxy" or
+    # "NO_PROXY" environment variables. To work around that, we check if
+    # the hostname is in the "no_proxy" or "NO_PROXY" environment variables
+    # ourselves using the "proxy_bypass()" function, which checks the "no_proxy"
+    # and "NO_PROXY" environment variables, and returns True if the host does
+    # not require a proxy server. The "lomond.WebSocket()" function will disable
+    # the proxy if the "proxies" parameter is an empty dictionary.  Otherwise,
+    # if the "proxies" parameter is "None", it will honor the "HTTP_PROXY" and
+    # "HTTPS_PROXY" environment variables. When the "proxies" parameter is not
+    # specified, the default value is "None".
+    proxies = {} if urllib.request.proxy_bypass(parsed_master.hostname) else None  # type: ignore
+
     url = request.make_url(master, "proxy/{}/".format(service))
-    ws = lomond.WebSocket(request.maybe_upgrade_ws_scheme(url))
+    ws = lomond.WebSocket(request.maybe_upgrade_ws_scheme(url), proxies=proxies)
```

Comment:
there's another lomond.WebSocket init below in line 188, and I suspect it'd also need the same handling?

Reply:
> there's another lomond.WebSocket init below in line 188, and I suspect it'd also need the same handling?

I was looking at the documentation in https://docs.determined.ai/latest/interfaces/proxy-ports.html

I ran those commands, but those commands alone aren't enough to exercise the code where the 2nd "lomond.WebSocket()" is.  What else do I need to do to force that code path?

Reply:
both `det e create config_file.yaml model_def -f -p 8265` and `det -m determined.cli.tunnel --listener 8265 $DET_MASTER $TASK_ID:8265` from the guide should exercise the second path.

Reply:
@ioga This is what I'm testing with.

I have the configuration file, as per the instructions in https://docs.determined.ai/latest/interfaces/proxy-ports.html

```
corujor@o184i054:~$ cat myconfig.yaml
environment:
  proxy_ports:
    - proxy_port: 8265
      proxy_tcp: true
```

I start a shell with that configuration file.

```
corujor@o184i054:/cstor/corujor/determined-ee$ DET_MASTER=http://localhost:8080 HOME=/cstor/corujor PYTHONPATH=/cstor/corujor/determined-ee/harness det shell start --config-file $HOME/myconfig.yaml
Launched shell (id: 498fe8d4-42a6-4ee2-b3c5-4a6e0e311ab9).
shell (id: 498fe8d4-42a6-4ee2-b3c5-4a6e0e311ab9) is ready.                      
Warning: Permanently added '498fe8d4-42a6-4ee2-b3c5-4a6e0e311ab9' (RSA) to the list of known hosts.
$ 
```

I put a print statement in the listener.

```
def _http_tunnel_listener(
    master_addr: str,
    tunnel: ListenerConfig,
    cert_file: Union[str, bool, None],
    cert_name: Optional[str],
    authorization_token: Optional[str] = None,
) -> socketserver.ThreadingTCPServer:
    parsed_master = request.parse_master_address(master_addr)
    assert parsed_master.hostname is not None, "Failed to parse master address: {}".format(
        master_addr
    )

    url = request.make_url(master_addr, "proxy/{}/".format(tunnel.service_id))

    class TunnelHandler(socketserver.BaseRequestHandler):
        def handle(self) -> None:

            print("RIGOBERTO:handle")

            ws = lomond.WebSocket(request.maybe_upgrade_ws_scheme(url)
```

I use the CLI to start a tunnel.

```
corujor@o184i054:/cstor/corujor/determined-ee$ MASTER=http://localhost:8080 PYTHONPATH=/cstor/corujor/determined-ee/harness python3 -m determined.cli.tunnel --listener 8265 --auth http://localhost:8080 498fe8d4-42a6-4ee2-b3c5-4a6e0e311ab9:8265
```

Nothing happens.  That is, my "print" statement doesn't show up.

I then ssh to port 8265, just to try "something" to get that path to get exercised.

```
corujor@o184i054:~$ ssh -p 8265 localhost
kex_exchange_identification: read: Connection reset by peer
Connection reset by 127.0.0.1 port 8265
```

That causes my "print" statement to show up and we get the "proxy fail" error.

```
corujor@o184i054:/cstor/corujor/determined-ee$ MASTER=http://localhost:8080 PYTHONPATH=/cstor/corujor/determined-ee/harness python3 -m determined.cli.tunnel --listener 8265 --auth http://localhost:8080 498fe8d4-42a6-4ee2-b3c5-4a6e0e311ab9:8265
RIGOBERTO:handle
error connecting to ws://localhost:8080/proxy/498fe8d4-42a6-4ee2-b3c5-4a6e0e311ab9:8265/; ('proxy fail; {} {}', 503, 'Service Unavailable')
Exception in thread Thread-4 (copy_from_websocket2):
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/cstor/corujor/determined-ee/harness/determined/cli/proxy.py", line 133, in copy_from_websocket2
    raise Exception("Connection failed: {}".format(event))
Exception: Connection failed: ConnectFail(reason='('proxy fail; {} {}', 503, 'Service Unavailable')')
```

I then modify the "_http_tunnel_listener" to ignore the proxies, as to avoid the "proxy fail" error.

```
def _http_tunnel_listener(
    master_addr: str,
    tunnel: ListenerConfig,
    cert_file: Union[str, bool, None],
    cert_name: Optional[str],
    authorization_token: Optional[str] = None,
) -> socketserver.ThreadingTCPServer:
    parsed_master = request.parse_master_address(master_addr)
    assert parsed_master.hostname is not None, "Failed to parse master address: {}".format(
        master_addr
    )

    url = request.make_url(master_addr, "proxy/{}/".format(tunnel.service_id))

    class TunnelHandler(socketserver.BaseRequestHandler):
        def handle(self) -> None:

            print("RIGOBERTO:handle")

            proxies = {} if urllib.request.proxy_bypass(parsed_master.hostname) else None  # type: ignore

            ws = lomond.WebSocket(request.maybe_upgrade_ws_scheme(url), proxies=proxies)

         ...
```

Then I ssh to port 8265 once again.

```
corujor@o184i054:~$ ssh -p 8265 localhost
kex_exchange_identification: read: Connection reset by peer
Connection reset by 127.0.0.1 port 8265
```

This time I get a different "Bad Gateway" error.

```
corujor@o184i054:/cstor/corujor/determined-ee$ MASTER=http://localhost:8080 PYTHONPATH=/cstor/corujor/determined-ee/harness python3 -m determined.cli.tunnel --listener 8265 --auth http://localhost:8080 498fe8d4-42a6-4ee2-b3c5-4a6e0e311ab9:8265
RIGOBERTO:handle
<response HTTP/1.1 502 Bad Gateway>
disconnecting websocket
Exception in thread Thread-4 (copy_from_websocket2):
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/cstor/corujor/determined-ee/harness/determined/cli/proxy.py", line 133, in copy_from_websocket2
    raise Exception("Connection failed: {}".format(event))
Exception: Connection failed: Rejected(response=<response HTTP/1.1 502 Bad Gateway>, reason='Websocket upgrade failed (code=502)')
```

I'm not sure if I'm getting the "Bad Gateway" error because ssh'ing to port 8265 is not the correct thing to do.

Other than ssh'ing to port 8265, what is the proper way to get this path to be exercised?

Reply:
do you have anything listening on the port 8265 inside the task?